pkg/image/registry/imagestreamtag/instantiate.go

```diff
@@ -0,0 +1,809 @@
+		return nil, errors.NewConflict(h.r.rest.gr, h.name, fmt.Errorf("the precondition UID does not match the image stream tag instantiate UID"))
+	}
+
+	return h.r.rest.completeInstantiate(h.ctx, tag, target, toImageStreamTagInstantiate(h.name, imageInstantiate, target), part, layerContentType)
+}
+
+func uploadLayer(ctx apirequest.Context, r io.Reader, repo distribution.Repository, mediaType string) (distribution.Descriptor, digest.Digest, *time.Time, error) {
+	if len(mediaType) == 0 {
+		return distribution.Descriptor{}, "", nil, fmt.Errorf("no content type set for layer")
+	}
+	blobs := repo.Blobs(ctx)
+	bw, err := blobs.Create(ctx)
+	if err != nil {
+		return distribution.Descriptor{}, "", nil, fmt.Errorf("unable to begin uploading blob: %v", err)
+	}
+
+	// calculate the blob digest as the sha256 sum of the uploaded contents
+	blobhash := sha256.New()
+
+	// calculate the diffID as the sha256 sum of the layer contents
+	pr, pw := io.Pipe()
+	layerhash := sha256.New()
+	ch := make(chan error)
+	var modTime *time.Time
+	go func() {
+		defer close(ch)
+		gr, err := gzip.NewReader(pr)
+		if err != nil {
+			ch <- fmt.Errorf("unable to create gzip reader layer upload: %v", err)
+			return
+		}
+		if !gr.Header.ModTime.IsZero() {
+			modTime = &gr.Header.ModTime
+		}
+		_, err = io.Copy(layerhash, gr)
+		ch <- err
+	}()
+
+	// upload the bytes
+	n, err := bw.ReadFrom(io.TeeReader(r, io.MultiWriter(blobhash, pw)))
+	if err != nil {
+		bw.Cancel(ctx)
+		return distribution.Descriptor{}, "", nil, fmt.Errorf("unable to upload new layer (%d): %v", n, err)
+	}
+	if err := pw.Close(); err != nil {
+		bw.Cancel(ctx)
+		return distribution.Descriptor{}, "", nil, fmt.Errorf("unable to complete writing diffID: %v", err)
+	}
+	if err := <-ch; err != nil {
+		bw.Cancel(ctx)
+		return distribution.Descriptor{}, "", nil, fmt.Errorf("unable to calculate layer diffID: %v", err)
+	}
+
+	// TODO: use configurable digests
+	layerdigest := digest.NewDigestFromBytes(digest.SHA256, layerhash.Sum(make([]byte, 0, layerhash.Size())))
+	blobdigest := digest.NewDigestFromBytes(digest.SHA256, blobhash.Sum(make([]byte, 0, blobhash.Size())))
+
+	glog.V(4).Infof("Wrote blob of %d bytes with digests layer=%s blob=%s", n, layerdigest, blobdigest)
+
+	desc, err := bw.Commit(ctx, distribution.Descriptor{
+		MediaType: mediaType,
+		Size:      n,
+		Digest:    blobdigest,
+	})
+	if err != nil {
+		bw.Cancel(ctx)
+		return distribution.Descriptor{}, "", nil, fmt.Errorf("unable to commit new layer: %v", err)
+	}
+	return desc, layerdigest, modTime, nil
+}
+
+// gzippedEmptyLayer is a gzip-compressed version of an empty tar file (1024 NULL bytes)
+// This comes from github.com/docker/distribution/manifest/schema1/config_builder.go; there is
+// a non-zero embedded timestamp; we could zero that, but that would just waste storage space
+// in registries, so let’s use the same values.
+var gzippedEmptyLayer = []byte{
```

Comment:
We already have it in few places. Maybe better to make it as a separate module ?

https://github.com/openshift/origin/search?utf8=%E2%9C%93&q=gzippedEmptyTar